### PR TITLE
provide wxXmlResourceHandler for wxVListBox

### LINKED ISSUE #24361: missing wxVListBoxXmlHandler?
I want to create a (subclass of) `wxVListBox` using XRC.  I initially thought that I could treat `wxVListBox` as a subclass of `wxPanel`, but then I discovered that `wxVListBox` has its own `Create()` function.  I think having its own `Create()` function means `wxVListBox` needs a `wxVListBoxXmlHandler` to properly be used with XRC.  

Assuming no one explains to me how my thinking here is incorrect, I will try to submit a `wxVListBoxXmlHandler`.

### PR DESCRIPTION
**Pull request by wsu-cb**
Fix #24361

### CODE REVIEW

**Review by vadz — COMMENTED** [review]
Looks good to me, thanks!

I wonder if we may want to add an attribute for the initial number of items? In practice it probably will never be fixed, as in the sample, though...

**Comment by wsu-cb** [comment]
> Looks good to me, thanks!
> 
You're welcome

> I wonder if we may want to add an attribute for the initial number of items? In practice it probably will never be fixed, as in the sample, though...

Yes, I didn't think there was much point in a count attribute.

File: src/xrc/xmlrsall.cpp [hunk 1]
```diff
@@ -128,6 +128,7 @@ void wxXmlResource::InitAllHandlers()
 #endif
 #if wxUSE_LISTBOX
     AddHandler(new wxListBoxXmlHandler);
+    AddHandler(new wxVListBoxXmlHandler);
```
[the remark this answers is not in the document]
Comment (wsu-cb) [reply src/xrc/xmlrsall.cpp:131]: Are you sure `wxVListBox` is always available?  `vlbox.cpp` has `#if wxUSE_LISTBOX` around the entire implementation of `wxVListBox`.  Or should that guard be removed?

File: samples/xrc/myframe.cpp [hunk 2]
```diff
@@ -449,3 +449,52 @@ void MyFrame::OnInfoBarShowMessage(wxCommandEvent& event)
...
 }
+
+#include <wx/vlbox.h>
+
+class wxVListBoxDerived : public wxVListBox
+{
+public:
+    // the derived class must implement this function to actually draw the item
+    // with the given index on the provided DC
+    virtual void OnDrawItem(wxDC& dc, const wxRect& rect, size_t n) const;
```
[the remark this answers is not in the document]
Comment (wsu-cb) [reply samples/xrc/myframe.cpp]: done

File: samples/xrc/myframe.cpp [hunk 3]
```diff
@@ -449,3 +449,52 @@ void MyFrame::OnInfoBarShowMessage(wxCommandEvent& event)
...
+
+    // the derived class must implement this method to return the height of the
+    // specified item
+    virtual wxCoord OnMeasureItem(size_t n) const;
```
[the remark this answers is not in the document]
Comment (wsu-cb) [reply samples/xrc/myframe.cpp]: done

File: samples/xrc/myframe.cpp [hunk 4]
```diff
@@ -449,3 +449,52 @@ void MyFrame::OnInfoBarShowMessage(wxCommandEvent& event)
...
+
+private:
+    wxDECLARE_DYNAMIC_CLASS(CTileListBoxWx);
+    wxDECLARE_EVENT_TABLE();
+
+    void OnCreate(wxWindowCreateEvent& event);
```
[the remark this answers is not in the document]
Comment (wsu-cb) [reply samples/xrc/myframe.cpp:474]: done

File: src/xrc/xh_vlistbox.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+wxIMPLEMENT_DYNAMIC_CLASS(wxVListBoxXmlHandler, wxXmlResourceHandler);
+
+wxVListBoxXmlHandler::wxVListBoxXmlHandler() : wxXmlResourceHandler()
+{
+    // panel styles
+    XRC_ADD_STYLE(wxTAB_TRAVERSAL);
+    /* I don't think it makes much sense for wxVListBox to have children
+    XRC_ADD_STYLE(wxWS_EX_VALIDATE_RECURSIVELY);
+    */
```
[the remark this answers is not in the document]
Comment (wsu-cb) [reply src/xrc/xh_vlistbox.cpp]: done

File: src/xrc/xh_vlistbox.cpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+    // listbox styles
+    /* docs say the only special style is */
```
[the remark this answers is not in the document]
Comment (wsu-cb) [reply src/xrc/xh_vlistbox.cpp]: done

File: src/xrc/xh_vlistbox.cpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+wxObject * wxVListBoxXmlHandler::DoCreateResource()
+{
+    /* non-standard because wxVListBox is an abstract
+        class, so "subclass" must be used */
+    wxVListBox *vlistbox = nullptr;
+    wxCHECK_MSG(m_instance,
+                nullptr,
+                "wxVListBox requires \"subclass\" attribute");
+    vlistbox = wxStaticCast(m_instance, wxVListBox);
```
[the remark this answers is not in the document]
Comment (wsu-cb) [reply src/xrc/xh_vlistbox.cpp]: done

**Comment by wsu-cb** [comment]
I tried to run `bakefile` to update the build files, but some of the build files didn't seem to change, and there are corresponding build failures. I don't know what to do about that.

**Comment by wsu-cb** [comment]
@vadz I made some comments and pushes, and request a re-review.

File: src/xrc/xmlrsall.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (vadz) [reply src/xrc/xmlrsall.cpp:131]: I've looked at the header which doesn't have `wxUSE_LISTBOX` check, but you're right, the implementation does, so let's keep it, sorry for a false alarm.

**Comment by vadz** [comment]
I've updated the MSVS project file, but `build/osx/wxcocoa.xcodeproj/project.pbxproj` still needs to be updated to add the new file (and maybe `build/osx/wxiphone.xcodeproj/project.pbxproj` too? I don't even know if we use XRC under iOS).

**Comment by wsu-cb** [comment]
> I've updated the MSVS project file, but `build/osx/wxcocoa.xcodeproj/project.pbxproj` still needs to be updated to add the new file (and maybe `build/osx/wxiphone.xcodeproj/project.pbxproj` too? I don't even know if we use XRC under iOS).

Thank you for fixing MSVS.  Judging by the build check results, only the Cocoa build still needs work.

**Comment by vadz** [comment]
I've run `build/osx/makeprojects/py`. It made a lot of changes, for some reason, and then I had to merge with master to avoid conflicts, let's see if it builds with Xcode now.

**Comment by wsu-cb** [comment]
> I've run `build/osx/makeprojects/py`. It made a lot of changes, for some reason, and then I had to merge with master to avoid conflicts, let's see if it builds with Xcode now.

All build check passed!  Thank you.

I have therefore marked this PR as "Ready for review".

**Pull request closed without merging**